enterprise/server/scim/scim.go

```diff
@@ -29,68 +99,236 @@ func NewSCIMServer(env environment.Env) *SCIMServer {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		switch r.Method {
-		case "GET":
-			s.getGroup(w, r)
-		case "POST":
-			s.createGroup(w, r)
-		case "PUT":
-			s.updateGroup(w, r)
-		case "PATCH":
-			s.updateGroup(w, r)
-		case "DELETE":
-			s.deleteGroup(w, r)
+	users := []*UserResource{}
+	filter := r.URL.Query().Get("filter")
+	if filter == "" {
+		displayUsers, err := s.env.GetUserDB().GetGroupUsers(ctx, g.GroupID, []grpb.GroupMembershipStatus{grpb.GroupMembershipStatus_MEMBER})
+		if err != nil {
+			return nil, err
+		}
+		for _, du := range displayUsers {
+			u := &tables.User{
+				UserID:    du.GetUser().GetUserId().GetId(),
+				FirstName: du.GetUser().GetName().GetFirst(),
+				LastName:  du.GetUser().GetName().GetLast(),
+				Email:     du.GetUser().GetEmail(),
+			}
+			users = append(users, newUserResource(u))
+		}
+	} else {
+		fu, err := s.getFilteredUsers(ctx, filter)
+		if err != nil {
+			return nil, err
 		}
+		users = fu
+	}
+	slices.SortFunc(users, func(a, b *UserResource) int {
+		return strings.Compare(a.UserName, b.UserName)
 	})
+	totalResults := len(users)
+
+	if startIndex > len(users) {
+		startIndex = len(users)
+	}
+	users = users[startIndex:]
+
+	if count == 0 {
+		count = len(users)
+	}
+	if count > len(users) {
+		count = len(users)
+	}
+	users = users[:count]
+
+	return &ListResponseResource{
+		Schemas:      []string{ListResponseSchema},
+		TotalResults: totalResults,
+		StartIndex:   startIndex + 1,
+		ItemsPerPage: count,
+		Resources:    users,
+	}, nil
 }
 
-func (s *SCIMServer) getGroup(w http.ResponseWriter, r *http.Request) {
-	// TODO
+func (s *SCIMServer) getUser(ctx context.Context, r *http.Request, g *tables.Group) (interface{}, error) {
+	id := path.Base(r.URL.Path)
+	u, err := s.env.GetUserDB().GetUserByID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return newUserResource(u), nil
 }
 
-func (s *SCIMServer) createGroup(w http.ResponseWriter, r *http.Request) {
-	// TODO
+func (s *SCIMServer) createUser(ctx context.Context, r *http.Request, g *tables.Group) (interface{}, error) {
+	req, err := io.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+	ur := UserResource{}
+	if err := json.Unmarshal(req, &ur); err != nil {
+		return nil, err
+	}
+	emailParts := strings.Split(ur.UserName, "@")
+	if len(emailParts) != 2 {
+		return nil, status.InvalidArgumentErrorf("invalid username %q", ur.UserName)
+	}
+	if emailParts[1] != g.OwnedDomain {
+		return nil, status.InvalidArgumentErrorf("username domain %q does not match group domain %q", ur.UserName, g.OwnedDomain)
+	}
+
+	pk, err := tables.PrimaryKeyForTable("Users")
+	if err != nil {
+		return nil, err
+	}
+	entityURL := build_buddy_url.WithPath("saml/metadata?slug=" + *g.URLIdentifier)
+	u := &tables.User{
+		UserID:    pk,
+		SubID:     fmt.Sprintf("%s/%s", entityURL, ur.UserName),
+		FirstName: ur.Name.GivenName,
+		LastName:  ur.Name.FamilyName,
+		Email:     ur.UserName,
+		Groups: []*tables.GroupRole{
```

Comment:
Is there a way to specify the User's role?

Reply:
this will need to be part of group management. theoretically you could have a user attribute that indicates the buildbuddy role but that feels gross.

Reply:
Sounds good